### PR TITLE
fix FileSystemWatcher_Unix_DoesNotLeak test

### LINKED ISSUE #2185: System.IO.FileSystem.Watcher.Tests failing on Libraries OSX Outerloop CI runs
There is no explicit failure listed in the logs. Possible these are just timing out. 

```
Executed on dci-mac-build-059
+ ./RunTests.sh --runtime-path /tmp/helix/working/BCEB09AF/p
----- start Tue Jan 21 03:39:10 PST 2020 =============== To repro directly: =====================================================
pushd .
/tmp/helix/working/BCEB09AF/p/dotnet exec --runtimeconfig System.IO.FileSystem.Watcher.Tests.runtimeconfig.json --depsfile System.IO.FileSystem.Watcher.Tests.deps.json xunit.console.dll System.IO.FileSystem.Watcher.Tests.dll -xml testResults.xml -nologo -nocolor -trait category=OuterLoop -notrait category=IgnoreForCI -notrait category=failing -notrait category=nonnetcoreapptests -notrait category=nonosxtests 
popd
===========================================================================================================
/private/tmp/helix/working/BCEB09AF/w/B274097E/e /private/tmp/helix/working/BCEB09AF/w/B274097E/e
  Discovering: System.IO.FileSystem.Watcher.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.IO.FileSystem.Watcher.Tests (found 19 of 184 test cases)
  Starting:    System.IO.FileSystem.Watcher.Tests (parallel test collections = on, max threads = 4)
   System.IO.FileSystem.Watcher.Tests: [Long Running Test] 'System.IO.Tests.DangerousFileSystemWatcherTests.FileSystemWatcher_Unix_DoesNotLeak', Elapsed: 00:02:11
   System.IO.FileSystem.Watcher.Tests: [Long Running Test] 'System.IO.Tests.DangerousFileSystemWatcherTests.FileSystemWatcher_Unix_DoesNotLeak', Elapsed: 00:04:12
   System.IO.FileSystem.Watcher.Tests: [Long Running Test] 'System.IO.Tests.DangerousFileSystemWatcherTests.FileSystemWatcher_Unix_DoesNotLeak', Elapsed: 00:06:13
   System.IO.FileSystem.Watcher.Tests: [Long Running Test] 'System.IO.Tests.DangerousFileSystemWatcherTests.FileSystemWatcher_Unix_DoesNotLeak', Elapsed: 00:08:13
   System.IO.FileSystem.Watcher.Tests: [Long Running Test] 'System.IO.Tests.DangerousFileSystemWatcherTests.FileSystemWatcher_Unix_DoesNotLeak', Elapsed: 00:10:14
   System.IO.FileSystem.Watcher.Tests: [Long Running Test] 'System.IO.Tests.DangerousFileSystemWatcherTests.FileSystemWatcher_Unix_DoesNotLeak', Elapsed: 00:12:15
   System.IO.FileSystem.Watcher.Tests: [Long Running Test] 'System.IO.Tests.DangerousFileSystemWatcherTests.FileSystemWatcher_Unix_DoesNotLeak', Elapsed: 00:14:16

```

Builds
- https://dev.azure.com/dnceng/public/_build/results?buildId=493872
- https://dev.azure.com/dnceng/public/_build/results?buildId=491012
- https://dev.azure.com/dnceng/public/_build/results?buildId=490329

Test Runs
- netcoreapp5.0-OSX-Release-x64-OSX.1013.Amd64.Open
- netcoreapp5.0-OSX-Release-x64-OSX.1013.Amd64.Open
- netcoreapp5.0-OSX-Release-x64-OSX.1014.Amd64.Open

### PR DESCRIPTION
**Pull request by wfurt**
FileSystemWatcher_Unix_DoesNotLeak tried to test the leak by running as many iterations needed to exhaust OS file descriptor limit in case of a leak. That can take a long time as well as it is then dependent on CI configuration. It seems like when the limit was increated recently that made this test super slow and timing out/killed in 15 minutes.   

Since no special privilege is needed to _decrese_ current limit, this change uses `RemoteExecutor` to spawn a new process, setting limit very low and running predictable number of iterations. It can now consistently finish in ~ 2 minutes regardless of the OS limit. 
(assuming it is at least 200) 

While working on the test I rolled back fix for #40888 and I verified the test fails. 
Note that when a process runs out of fds it can fail in unusual ways - like claiming it runs out of memory or failing to load type. 

fixes  #2185

### CODE REVIEW

**Comment by wfurt** [comment]
All the relevant Outerloop test failed for unrelated reasons (System.Threading.Tasks) 
Updated test passed on OSX and all Linux versions. Still failing a loot in master.

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs [hunk 1]
```diff
@@ -10,47 +10,77 @@
...
 using System.Threading;
 using Microsoft.DotNet.XUnitExtensions;
+using Microsoft.DotNet.RemoteExecutor;
 using Xunit;
 using Xunit.Abstractions;
 
 namespace System.IO.Tests
 {
-    public partial class DangerousFileSystemWatcherTests
+    // Helper class to avoid issues with RemoteExecutor and _output and other properties.
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: What issues and how is this working around them?  I'm not seeing it.

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs [hunk 2]
```diff
@@ -10,47 +10,77 @@
...
-            _output.WriteLine("File descriptor limit is {0}", limits.CurrentLimit);
+            _count = count;
+            _limit = limit;
+            _path = path;
+        }
 
-            if (limits.CurrentLimit > 50_000)
+        internal void ExecuteTest()
+        {
+            RemoteInvokeOptions options = new RemoteInvokeOptions{ TimeOut = 600_000 };
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: Nit:
```suggestion
            RemoteInvokeOptions options = new RemoteInvokeOptions { TimeOut = 600_000 };
```

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs [hunk 3]
```diff
@@ -10,47 +10,77 @@
...
+            RemoteExecutor.Invoke((countString, openFileLimitString, testDirectory) =>
             {
-                throw new SkipTestException($"File descriptor limit is too high {limits.CurrentLimit}.");
-            }
+                ulong maxFd = ulong.Parse(openFileLimitString);
+                int count = Int32.Parse(countString);
+                Interop.Sys.RLimit limits = new Interop.Sys.RLimit{ CurrentLimit = maxFd, MaximumLimit = maxFd};
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]:
```suggestion
                Interop.Sys.RLimit limits = new Interop.Sys.RLimit { CurrentLimit = maxFd, MaximumLimit = maxFd};
```

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs [hunk 4]
```diff
@@ -10,47 +10,77 @@
...
-                    using (var watcher = new FileSystemWatcher(TestDirectory))
+                    for (int i = 0; i < count; i++)
                     {
-                        watcher.Created += (s, e) => { } ;
-                        watcher.EnableRaisingEvents = true;
+                        using (var watcher = new FileSystemWatcher(testDirectory))
+                        {
+                            watcher.Created += (s, e) => { } ;
+                            watcher.EnableRaisingEvents = true;
+                            watcher.EnableRaisingEvents = false;
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: Is this necessary?

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs (hunk 1 above)
Comment (wfurt) [reply src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: I'm getting 
```
     System.IO.Tests.DangerousFileSystemWatcherTests.FileSystemWatcher_Unix_DoesNotLeak [FAIL]
        Assert.All() Failure: 1 out of 1 items in the collection did not pass.
        [0]: Item: Xunit.Abstractions.ITestOutputHelper _output
             Xunit.Sdk.TrueException: Field marshaling is not supported by Invoke: _output
             Expected: True
             Actual:   False
                at Xunit.Assert.True(Nullable`1 condition, String userMessage) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\BooleanAsserts.cs:line 96
                at Xunit.Assert.True(Boolean condition, String userMessage) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\BooleanAsserts.cs:line 84
                at Microsoft.DotNet.RemoteExecutor.RemoteExecutor.<>c.<GetMethodInfo>b__24_0(FieldInfo fi) in /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs:line 351
                at Xunit.Assert.All[T](IEnumerable`1 collection, Action`1 action) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\CollectionAsserts.cs:line 36
        Stack Trace:
          /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs(354,0): at Microsoft.DotNet.RemoteExecutor.RemoteExecutor.GetMethodInfo(Delegate d)
          /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs(212,0): at Microsoft.DotNet.RemoteExecutor.RemoteExecutor.Invoke(Func`2 method, String arg, RemoteInvokeOptions options)
          /Users/furt/github/wfurt-runtime/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs(89,0): at System.IO.Tests.DangerousFileSystemWatcherTests.FileSystemWatcher_Unix_DoesNotLeak()
```

and `_output` is _not_  used inside of the remote code. I had a prototype where I ditched _outout completely and it worked but I somewhat like the ability to instrument. 

I don't know if there is any better way how to get around this.

Comment (stephentoub) [reply src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: You're getting that with what code?

Comment (wfurt) [reply src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: when I move the body of `ExecuteTest()` to the `FileSystemWatcher_Unix_DoesNotLeak()`  itself. For that reason, I moved the test body to separate class to defeat checks from `GetMethodInfo()`

Comment (wfurt) [reply src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: I the issue and it is working now without extra class.

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs (hunk 4 above)
Comment (wfurt) [reply src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: no, it is not. I had issues reproducing the original problem and failing the test. I don't know if it makes sense to test disposal in both cases when the watcher is running or not. But for now, I removed the last assignment to keep it simple.

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs [hunk 5]
```diff
@@ -23,34 +24,43 @@ public void FileSystemWatcher_Unix_DoesNotLeak()
         {
             Interop.Sys.GetRLimit(Interop.Sys.RlimitResources.RLIMIT_NOFILE, out Interop.Sys.RLimit limits);
             _output.WriteLine("File descriptor limit is {0}", limits.CurrentLimit);
+            _output.WriteLine($"Starting 100/200 test on {TestDirectory}");
 
-            if (limits.CurrentLimit > 50_000)
+            RemoteInvokeOptions options = new RemoteInvokeOptions { TimeOut = 600_000 };
+            RemoteExecutor.Invoke((testDirectory) =>
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: Nit:
```suggestion
            RemoteExecutor.Invoke(testDirectory =>
```

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs [hunk 6]
```diff
@@ -23,34 +24,43 @@ public void FileSystemWatcher_Unix_DoesNotLeak()
...
-                throw;
-            }
+                catch (Exception e)
+                {
+                    // If we use all handles we may not have luck writing out errors.
+                    // Try our best here. _output is not available within RemoteExec().
+                    // When we run out of fd, exception may be weird from OutOfMem to "unable load type".
+                    Console.WriteLine($"Test failed for count={count}, limnit={maxFd} and path='{testDirectory}'.");
+                    Console.WriteLine(e.Message);
+                    return 1;
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: Why not `throw;` instead of `return 1;`?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs [hunk 7]
```diff
@@ -23,34 +24,43 @@ public void FileSystemWatcher_Unix_DoesNotLeak()
...
-                Console.WriteLine(e.Message);
-                _output.WriteLine(e.Message);
-                throw;
-            }
+                catch (Exception e)
+                {
+                    // If we use all handles we may not have luck writing out errors.
+                    // Try our best here. _output is not available within RemoteExec().
+                    // When we run out of fd, exception may be weird from OutOfMem to "unable load type".
+                    Console.WriteLine($"Test failed for count={count}, limnit={maxFd} and path='{testDirectory}'.");
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]:
```suggestion
                    Console.WriteLine($"Test failed for count={count}, limit={maxFd} and path='{testDirectory}'.");
```

File: src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs (hunk 6 above)
Comment (wfurt) [reply src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]: When you throw when there are no file descriptors, I was getting some weird exception about loading types and the stack trace did not work either. We are basically at the point the code can not function properly.

**Comment by wfurt** [comment]
OSX failed only in System.Net.Http.Functional.Tests.

**Pull request merged**